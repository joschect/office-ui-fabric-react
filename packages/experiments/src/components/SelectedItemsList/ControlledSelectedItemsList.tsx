import * as React from 'react';

import { IControlledSelectedItemsList, IControlledSelectedItemListProps, BaseSelectedItem } from './SelectedItemsList.types';
import { copyToClipboard } from '../../utilities/copyToClipboard';
import { useSelectionIndices } from '../../utilities/useSelectionIndices';

/**
 * An externally-managed selected item list.
 */
export const ControlledSelectedItemsList = <TItem extends BaseSelectedItem>(props: IControlledSelectedItemListProps<TItem>) => {
  const selection = React.useMemo(() => props.selection, [props.selection]);
  const selectedIndices = useSelectionIndices(selection);
  // Synchronize the selection against the items in the selection
  React.useEffect(() => {
    selection &&
      selection.setItems(
        props.selectedItems,
        false // shouldClear
      );
  });

  const onRemoveItemCallbacks = React.useMemo(
    () =>
      // create callbacks ahead of time with memo.
      // (hooks have to be called in the same order)
      props.selectedItems.map((item: TItem) =>
        // do not generate callbacks for items that cannot be removed
        !props.canRemoveItem || props.canRemoveItem(item) ? () => props.onItemsRemoved([item]) : undefined
      ),
    [props.selectedItems, props.canRemoveItem, props.onItemsRemoved]
  );

  // only used in the imperative handle
  const copyItemsInSelectionToClipboard = React.useCallback((): void => {
    if (props.getItemCopyText && selectedIndices.length > 0) {
      const itemsInSelection = selectedIndices.map(itemIndex => props.selectedItems[itemIndex]);
      const copyText = props.getItemCopyText(itemsInSelection);
      copyToClipboard(copyText);
    }
  }, [props.selectedItems, props.getItemCopyText, selectedIndices]);

  React.useImperativeHandle(
    props.componentRef,
    () => ({
      copyItemsInSelectionToClipboard
    }),
    [copyItemsInSelectionToClipboard]
  );

  const SelectedItem = props.onRenderItem;
  return (
    <>
      {props.selectedItems.map((item: TItem, index: number) => (
        <SelectedItem
          item={item}
          index={index}
          key={item.key !== undefined ? item.key : index}
          selected={selectedIndices.indexOf(index) !== -1}
          removeButtonAriaLabel={props.removeButtonAriaLabel}
          onRemoveItem={onRemoveItemCallbacks[index]}
          onItemChange={props.onItemChange}
        />
      ))}
    </>
  );
};
export type ControlledSelectedItemsList<TItem extends BaseSelectedItem> = (
  props: IControlledSelectedItemListProps<TItem> & React.RefAttributes<IControlledSelectedItemsList>
) => React.ReactElement;
(ControlledSelectedItemsList as any).displayName = 'ControlledSelectedItemsList';
