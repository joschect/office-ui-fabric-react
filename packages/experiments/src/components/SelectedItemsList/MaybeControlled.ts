export type MaybeControlled<TControlledProps, TUncontrolledProps> =
  | ({
      isControlled: true;
    } & TControlledProps)
  | {
      isControlled: false;
    } & TUncontrolledProps;

export type ControlledProps<T> = T extends MaybeControlled<infer C, infer U>
  ? (C & {
      isControlled: true;
    })
  : never;

export type UncontrolledProps<T> = T extends MaybeControlled<infer C, infer U>
  ? (U & {
      isControlled: false;
    })
  : never;
