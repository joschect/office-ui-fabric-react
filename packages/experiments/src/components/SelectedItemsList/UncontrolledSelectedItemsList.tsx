import * as React from 'react';
import { Selection } from 'office-ui-fabric-react/lib/Selection';

import {
  IUncontrolledSelectedItemsList,
  IControlledSelectedItemsList,
  IUncontrolledSelectedItemListProps,
  BaseSelectedItem
} from './SelectedItemsList.types';
import { useSelectionIndices } from '../../utilities/useSelectionIndices';
import { ControlledSelectedItemsList } from './ControlledSelectedItemsList';

export const useSelectedItemListState = <TSelectedItem extends any>(defaultSelectedItems: TSelectedItem[] | undefined) => {
  const [selectedItems, setSelectedItems] = React.useState(defaultSelectedItems || []);

  const removeSelectedItems = React.useCallback(
    (removedItems: TSelectedItem[]) => {
      // Update internal state if it is tracked at all
      const nextSelectedItems = selectedItems.filter(i => removedItems.indexOf(i) === -1);
      if (nextSelectedItems.length !== selectedItems.length - removedItems.length) {
        console.error('removeSelectedItems called on some item not in the current state');
      }
      setSelectedItems(nextSelectedItems);
    },
    [selectedItems]
  );

  const replaceSelectedItem = React.useCallback(
    (newItem: TSelectedItem | TSelectedItem[], index: number): void => {
      const newItemsArray = !Array.isArray(newItem) ? [newItem] : newItem;

      if (index >= 0) {
        const newItems: TSelectedItem[] = [...selectedItems];
        newItems.splice(index, 1, ...newItemsArray);
        setSelectedItems(newItems);
      }
    },
    [setSelectedItems, selectedItems]
  );

  const appendSelectedItems = React.useCallback(
    (itemsToAppend: TSelectedItem[]) => {
      setSelectedItems(selectedItems.concat(itemsToAppend));
    },
    [selectedItems]
  );

  return {
    selectedItems,
    setSelectedItems,
    removeSelectedItems,
    replaceSelectedItem,
    appendSelectedItems
  };
};

/**
 * A self-managed selected item list.
 */
export const UncontrolledSelectedItemsList = <TItem extends BaseSelectedItem>(props: IUncontrolledSelectedItemListProps<TItem>) => {
  const selection = React.useMemo(() => props.selection || new Selection(), [props.selection]);
  const selectedIndices = useSelectionIndices(selection);

  const { selectedItems, removeSelectedItems, replaceSelectedItem, appendSelectedItems } = useSelectedItemListState(
    props.defaultSelectedItems
  );

  // Selection which initializes at the beginning of the component and
  // only updates if selection becomes set in props (e.g component transitions from
  // being controlled to uncontrolled)
  const itemsInSelection = React.useMemo(() => selectedIndices.filter(i => i > 0 && i < selectedItems.length).map(i => selectedItems[i]), [
    selectedItems,
    selectedIndices
  ]);

  const unselectAll = React.useCallback(() => {
    selection.setAllSelected(false);
  }, [selectedItems]);

  const controlledRef = React.useRef<IControlledSelectedItemsList>(null);
  const copyItemsInSelectionToClipboard = React.useCallback(() => {
    controlledRef.current && controlledRef.current.copyItemsInSelectionToClipboard();
  }, [controlledRef]);

  // For usage as a controlled component with a ref
  React.useImperativeHandle(
    props.componentRef,
    (): IUncontrolledSelectedItemsList<TItem> => ({
      items: selectedItems,
      addItems: appendSelectedItems,
      removeItems: removeSelectedItems,
      itemsInSelection,
      unselectAll,
      copyItemsInSelectionToClipboard
    }),
    [selectedItems, itemsInSelection, appendSelectedItems, unselectAll, removeSelectedItems, copyItemsInSelectionToClipboard]
  );

  return (
    <ControlledSelectedItemsList<TItem>
      componentRef={controlledRef}
      // pass-through props
      getItemCopyText={props.getItemCopyText}
      onRenderItem={props.onRenderItem}
      removeButtonAriaLabel={props.removeButtonAriaLabel}
      canRemoveItem={props.canRemoveItem}
      // props that are internal to the uncontrolled component
      selection={selection}
      selectedItems={selectedItems}
      onItemsRemoved={removeSelectedItems}
      onItemChange={replaceSelectedItem}
    />
  );
};
export type UncontrolledSelectedItemsList<TItem extends BaseSelectedItem> = (
  props: IUncontrolledSelectedItemListProps<TItem>
) => React.ReactElement;
(UncontrolledSelectedItemsList as any).displayName = 'UncontrolledSelectedItemsList';
