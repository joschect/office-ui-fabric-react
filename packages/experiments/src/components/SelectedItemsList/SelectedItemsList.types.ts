import * as React from 'react';
import { IPickerItemProps } from 'office-ui-fabric-react/lib/Pickers';
import { Selection } from 'office-ui-fabric-react/lib/Selection';
import { MaybeControlled } from './MaybeControlled';

/**
 * Ref type of the UncontrolledSelectedItemList
 */
export interface IUncontrolledSelectedItemsList<T> {
  /**
   * Current value of the input
   */
  items: T[];
  /**
   * The items that are in the selection -- either the one passed in
   * via props, or the internal selection maintained by the component
   * when no selection is provided in the props
   */
  itemsInSelection: T[] | undefined;
  /*
   * Adds items to the selection
   */
  addItems: (items: T[]) => void;
  /*
   * Removes all items from the selection.
   * If called with a selection passed in, this will mutate the
   * selection in props. Prefer to update the selection yourself.
   */
  unselectAll: () => void;
  /**
   * Removes items from the selection
   */
  removeItems: (items: T[]) => void;
  /**
   * Copies the current selected items to the clipboard.
   */
  copyItemsInSelectionToClipboard: () => void;
}

/**
 * Ref type of the ControlledSelectedItemList
 */
export interface IControlledSelectedItemsList {
  /**
   * Copies the current selected items to the clipboard.
   */
  copyItemsInSelectionToClipboard: () => void;
}

export interface ISelectedItemProps<T> extends IPickerItemProps<T> {
  /**
   * Override onItemChange to support replacing an item with multiple items.
   */
  onItemChange: (newItem: T | T[], index: number) => void;
}

export type BaseSelectedItem = {
  key?: React.Key;
};

/**
 * Props that are constant on the selected item list when used as controlled or uncontrolled.
 */
export interface ICommonSelectedItemListProps<T> {
  /**
   * Gets the copy text that will be set in the item.
   */
  getItemCopyText?: (items: T[]) => string;
  /**
   * Function that specifies how the selected item will appear.
   */
  onRenderItem: React.ComponentType<ISelectedItemProps<T>>;
  /**
   * Aria label for the 'X' button in the selected item component.
   * @defaultValue ''
   */
  removeButtonAriaLabel?: string;

  /**
   * A callback on whether this item can be removed
   */
  canRemoveItem?: (item: T) => boolean;
}

/**
 * Legal props for the selected items list when used as a controlled (managed) component.
 */
export interface IControlledSelectedItemListProps<T> extends ICommonSelectedItemListProps<T> {
  componentRef?: React.Ref<IControlledSelectedItemsList>;
  /**
   * The selection
   */
  selection: Selection;
  /**
   * The items that the base picker should currently display as selected.
   * controlled component.
   *
   */
  selectedItems: T[];
  /**
   * A callback when and item or items are removed
   */
  onItemsRemoved: (removedItems: T[]) => void;
  /**
   * Override onItemChange to support replacing an item with multiple items.
   */
  onItemChange: (newItem: T | T[], index: number) => void;
}

/**
 * Legal props for the selected items list when used as an uncontrolled (self-managing) component
 */
export interface IUncontrolledSelectedItemListProps<T> extends ICommonSelectedItemListProps<T> {
  componentRef?: React.Ref<IUncontrolledSelectedItemsList<T>>;
  /**
   * The selection
   */
  selection?: Selection;
  /**
   * Initial items that have already been selected and should appear in the list.
   */
  defaultSelectedItems?: T[];
  /**
   * A callback for when the selected list of items changes.
   */
  onChange?: (items?: T[]) => void;
}

export type ISelectedItemsListProps<T> = MaybeControlled<IControlledSelectedItemListProps<T>, IUncontrolledSelectedItemListProps<T>>;
