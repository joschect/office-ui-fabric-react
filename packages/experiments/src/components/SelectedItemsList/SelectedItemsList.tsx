import * as React from 'react';

import { ISelectedItemsListProps, BaseSelectedItem } from './SelectedItemsList.types';
import { ControlledSelectedItemsList } from './ControlledSelectedItemsList';
import { UncontrolledSelectedItemsList } from './UncontrolledSelectedItemsList';

/**
 * Will render either a controlled or uncontrolled selected items depending on the props that are passed in.
 */
export const SelectedItemsList = React.memo(<TItem extends BaseSelectedItem>(props: ISelectedItemsListProps<TItem>) => {
  if (props.isControlled) {
    return <ControlledSelectedItemsList<TItem> {...props} />;
  } else {
    return <UncontrolledSelectedItemsList<TItem> {...props} />;
  }
  // Cast back to a generic function type, since typescript <3.4 collapses generics on higher-order functions
  // This was solved by https://github.com/microsoft/TypeScript/pull/30215 in typescript@3.4,
}) as <TItem extends BaseSelectedItem>(props: ISelectedItemsListProps<TItem>) => React.ReactElement;
export type SelectedItemsList<TItem extends BaseSelectedItem> = (props: ISelectedItemsListProps<TItem>) => React.ReactElement;
