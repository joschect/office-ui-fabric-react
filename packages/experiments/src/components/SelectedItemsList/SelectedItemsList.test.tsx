import * as React from 'react';
import { act, create, ReactTestRenderer } from 'react-test-renderer';

import { ISelectedItemProps, IUncontrolledSelectedItemsList } from './SelectedItemsList.types';
import { SelectedItemsList } from './SelectedItemsList';
import { Selection as FabricSelection } from 'office-ui-fabric-react/lib/Selection';

export interface ISimple {
  key: string;
  name: string;
}

const basicItemRenderer = (props: ISelectedItemProps<ISimple>) => {
  return <div key={props.item.key}> {props.item.name} </div>;
};

const doNothing = () => undefined;

describe('SelectedItemsList', () => {
  describe('snapshots', () => {
    it('renders correctly when empty', () => {
      const component = create(<SelectedItemsList<ISimple> isControlled={false} onRenderItem={basicItemRenderer} />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders correctly with children', () => {
      const listRef = React.createRef<IUncontrolledSelectedItemsList<ISimple>>();
      const component = create(<SelectedItemsList<ISimple> componentRef={listRef} isControlled={false} onRenderItem={basicItemRenderer} />);

      act(() => {
        listRef.current &&
          listRef.current.addItems([
            {
              key: 'key-1',
              name: 'Joey Wheeler'
            },
            {
              key: 'key-2',
              name: 'Tristan Taylor'
            }
          ]);
      });

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('can remove items', () => {
    // Arrange
    const onChange = (items: ISimple[] | undefined): void => {
      expect(items!.length).toBe(1);
      expect(items![0].name).toBe('b');
    };
    const listRef = React.createRef<IUncontrolledSelectedItemsList<ISimple>>();

    // Act
    act(() => {
      create(
        <SelectedItemsList<ISimple>
          defaultSelectedItems={[{ key: '1', name: 'a' }, { key: '2', name: 'b' }]}
          isControlled={false}
          componentRef={listRef}
          onRenderItem={basicItemRenderer}
          onChange={onChange}
        />
      );
    });

    // Assert
    if (!listRef.current) {
      throw new Error('listRef was not initialized');
    }
    expect(listRef.current.items.length).toEqual(2);
    act(() => {
      listRef.current && listRef.current.removeItems([listRef.current.items[1]]);
    });
    expect(listRef.current.items.length).toEqual(1);
  });

  it('can add items', () => {
    const listRef = React.createRef<IUncontrolledSelectedItemsList<ISimple>>();
    create(<SelectedItemsList<ISimple> isControlled={false} componentRef={listRef} onRenderItem={basicItemRenderer} />);

    const items: ISimple[] = [{ key: '1', name: 'a' }, { key: '2', name: 'b' }];
    if (!listRef.current) {
      throw new Error('listRef was not initialized');
    }

    act(() => {
      listRef.current && listRef.current.addItems(items);
    });
    expect(listRef.current.items.length).toEqual(2);
  });

  describe('selection', () => {
    let selection: FabricSelection;
    beforeEach(() => {
      selection = new FabricSelection();
    });

    const mountListWithData = (data: ISimple[]) => {
      // Run in 'act' to trigger effects after layout
      act(() => {
        create(
          <SelectedItemsList<ISimple>
            isControlled={true}
            selection={selection}
            onRenderItem={basicItemRenderer}
            selectedItems={data}
            onItemChange={doNothing}
            onItemsRemoved={doNothing}
          />
        );
      });
    };

    it('returns selected items via the selection', () => {
      mountListWithData([{ key: '1', name: 'a' }, { key: '2', name: 'b' }]);
      expect(selection.getItems()).toEqual([{ key: '1', name: 'a' }, { key: '2', name: 'b' }]);
    });

    it('can select items via the selection', () => {
      mountListWithData([{ key: '1', name: 'a' }, { key: '2', name: 'b' }]);
      act(() => {
        selection.setAllSelected(false);
        selection.setIndexSelected(
          0,
          true, // isSelected
          false // shouldAnchor
        );
      });
      expect(selection.getSelection()).toEqual([{ key: '1', name: 'a' }]);
    });

    it('updates items in selection when the selectedItems are updated by props', () => {
      let tree: ReactTestRenderer | undefined;
      act(() => {
        tree = create(
          <SelectedItemsList<ISimple>
            isControlled={true}
            selection={selection}
            onRenderItem={basicItemRenderer}
            selectedItems={[{ key: '1', name: 'a' }, { key: '2', name: 'b' }]}
            onItemChange={doNothing}
            onItemsRemoved={doNothing}
          />
        );
      });

      act(() => {
        tree!.update(
          <SelectedItemsList<ISimple>
            isControlled={true}
            selection={selection}
            onRenderItem={basicItemRenderer}
            selectedItems={[{ key: '1', name: 'alphonse' }, { key: '2', name: 'edward' }]}
            onItemChange={doNothing}
            onItemsRemoved={doNothing}
          />
        );
      });

      expect(selection.getItems()).toEqual([{ key: '1', name: 'alphonse' }, { key: '2', name: 'edward' }]);
    });
  });
});
