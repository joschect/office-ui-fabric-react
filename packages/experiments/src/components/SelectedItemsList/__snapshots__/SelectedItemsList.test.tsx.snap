// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedItemsList snapshots renders correctly when empty 1`] = `null`;

exports[`SelectedItemsList snapshots renders correctly with children 1`] = `
Array [
  <div>
     
    Joey Wheeler
     
  </div>,
  <div>
     
    Tristan Taylor
     
  </div>,
]
`;
